Check for empty Namespace

icnocop opened the pull request:
<!-- markdownlint-disable MD002 MD041 -->

### Issues

Fixes #952

### Description

Check if the `Namespace` is an empty string when returning the `FullName` of `EnumTypeConfiguration` and `StructuralTypeConfiguration`.

### Checklist (Uncheck if it is not completed)

- [x] *Test cases added*
- [x] *Build and test with one-click build and test script passed*

gathogojr commented:
@icnocop Based on [the specification](http://docs.oasis-open.org/odata/odata/v4.0/errata03/os/complete/part3-csdl/odata-v4.0-errata03-os-part3-csdl-complete.html#_Toc453752522), should the value for the namespace ever be null or empty?

icnocop commented:
The spec indicates the namespace must be unique.

It doesn't explicitly say it must not be empty.

Does the spec need to be updated?

Currently, if you set the namespace to an empty string, the 'FullName' will begin with a period, which is not expected.

If the spec needs to be updated to clarify that requirement, an exception should occur if the namespace is being set to an empty string.

mikepizzo commented:
@icnocop -- thanks for the contribution!

Clearly the existing logic is flawed, as it allows setting the namespace to an empty string and then builds a qualified name prefixed with ".".

From Section 6 of the [abnf](http://docs.oasis-open.org/odata/odata/v4.01/os/abnf/odata-abnf-construction-rules.txt):
qualifiedEnumTypeName = namespace "." enumerationTypeName
namespace  = namespacePart *( "." namespacePart )
namespacePart = odataIdentifier
; Note: this pattern is overly restrictive, the normative definition is type TSimpleIdentifier in OData EDM XML Schema
odataIdentifier = identifierLeadingCharacter *127identifierCharacter

Which all means that namespace must have at least one character.

So the real issue in the code is that it lets the user set the namespace to an empty string, which should not be allowed.

icnocop commented:
Thank you for referencing the ABNF.

I've updated the code to throw an `ArgumentException` if the `StructuralTypeConfiguration.Namespace` or `StructuralTypeConfiguration.Name` properties are set to null, empty, or consists of only white-space characters.

I reverted the changes I had made to check the same properties for `EnumTypeConfiguration` because it seems to behave differently.

Thank you.

icnocop commented:
Let me know if I should squash my commits into one.

mikepizzo commented:
Hi @icnocop  -- thanks for the follow-up. No need to squash, we'll do that when we merge.

Slight concern that this is a breaking change (i.e, if someone were previously setting namespace to null it wouldn't throw an exception), but given the fact that it would almost certainly break further down the line, I think it's probably safe to accept.

mikepizzo commented on test/UnitTest/Microsoft.AspNet.OData.Test.Shared/Builder/StructuralTypeConfigurationOfTStructuralTypeTest.cs:
```diff
@@ -4,6 +4,7 @@
 using Microsoft.AspNet.OData.Builder;
 using Microsoft.AspNet.OData.Test.Common;
 using Moq;
+using System;
```
Nit: System using statements should come before other using statements.

icnocop replied:
Fixed. Thank you.

mikepizzo reviewed (approved):
:shipit:

KenitoInc reviewed (approved):
LGTM

gathogojr commented on test/UnitTest/Microsoft.AspNet.OData.Test.Shared/Builder/StructuralTypeConfigurationTest.cs:
```diff
@@ -188,5 +188,39 @@ public void NamespaceAssignment_AutoAssignsNamespaceToStructuralType_NamespaceAs
         }
+
+        /// <summary>
+        /// Tests the Namespace property setter logic with a null or white space value throws an argument exception.
```
@icnocop Instead of having 3 asserts for 3 different scenarios, we could use:
```csharp
[Theory]
[InlineData(null)]
[InlineData("")
[InlineData(" ")]
public void NamespaceAssignment_WithNullOrWhiteSpace_ThrowsArgumentException(string namespace)
{
    // ...
    ExceptionAssert.ThrowsArgument(() => entityTypeConfiguration.Namespace = namespace, "value");
}
```

icnocop replied:
Fixed. Thank you.

gathogojr commented on test/UnitTest/Microsoft.AspNet.OData.Test.Shared/Builder/StructuralTypeConfigurationTest.cs:
```diff
@@ -188,5 +188,39 @@ public void NamespaceAssignment_AutoAssignsNamespaceToStructuralType_NamespaceAs
+        /// Tests the Name property setter logic with a null or white space value throws an argument exception.
+        /// </summary>
+        [Fact]
+        public void NameAssignment_WithNullOrWhiteSpace_ThrowsArgumentException()
```
@icnocop Instead of having 3 asserts for 3 different scenarios, we could use:
```csharp
[Theory]
[InlineData(null)]
[InlineData("")
[InlineData(" ")]
public void NameAssignment_WithNullOrWhiteSpace_ThrowsArgumentException(string name)
{
    // ...
    ExceptionAssert.ThrowsArgument(() => entityTypeConfiguration.Name = name, "value");
}
```

icnocop replied:
Fixed. Thank you.

gathogojr commented on test/UnitTest/Microsoft.AspNet.OData.Test.Shared/Builder/StructuralTypeConfigurationTest.cs:
```diff
@@ -188,5 +188,41 @@ public void NamespaceAssignment_AutoAssignsNamespaceToStructuralType_NamespaceAs
+        [InlineData(null)]
+        [InlineData("")]
+        [InlineData(" ")]
+        public void NameAssignment_WithNullOrWhiteSpace_ThrowsArgumentException(string @namespace)
```
@icnocop Maybe use `name` to avoid confusion...

icnocop replied:
Done. Thank you.
